pkg/kubectl/cmd/resource/get.go

```diff
@@ -320,145 +345,51 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
-				updatePrintOptionsForOpenAPI(f, mapping, printOpts)
-			}
-
-			printer, err = cmdutil.PrinterForOptions(printOpts)
-			if err != nil {
-				if !errs.Has(err.Error()) {
-					errs.Insert(err.Error())
-					allErrs = append(allErrs, err)
-				}
-				continue
-			}
-
-			// TODO: this doesn't belong here
-			// add linebreak between resource groups (if there is more than one)
-			// skip linebreak above first resource group
-			if lastMapping != nil && !noHeaders {
-				fmt.Fprintf(options.ErrOut, "%s\n", "")
-			}
 
-			lastMapping = mapping
-		}
+	// TODO(juanvallejo): handling this inside of humanprinter_flags.go causes an import cycle
+	// figure out how to attach these handlers without resorting to each command having to do it.
+	printersinternal.AddHandlers(printer.(*printers.HumanReadablePrinter))
 
-		typedObj := info.AsInternal()
-
-		if resourcePrinter, found := printer.(*printers.HumanReadablePrinter); found {
-			resourceName := resourcePrinter.GetResourceKind()
-			if mapping != nil {
-				if resourceName == "" {
-					resourceName = mapping.Resource
-				}
-				if alias, ok := kubectl.ResourceShortFormFor(mapping.Resource); ok {
-					resourceName = alias
-				} else if resourceName == "" {
-					resourceName = "none"
-				}
-			} else {
-				resourceName = "none"
-			}
+	list := api.List{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "List",
+			APIVersion: "v1",
+		},
+		ListMeta: metav1.ListMeta{},
+	}
+	for _, info := range infos {
+		list.Items = append(list.Items, info.AsInternal())
+	}
 
-			if showKind {
-				resourcePrinter.EnsurePrintWithKind(resourceName)
-			}
+	listData, err := json.Marshal(list)
+	if err != nil {
+		return err
+	}
 
-			if err := printer.PrintObj(typedObj, w); err != nil {
-				if !errs.Has(err.Error()) {
-					errs.Insert(err.Error())
-					allErrs = append(allErrs, err)
-				}
-			}
-			continue
-		}
-		objToPrint := typedObj
-		if printer.IsGeneric() {
-			// use raw object as received from the builder when using generic
-			// printer instead of decodedObj
-			objToPrint = original
-		}
-		if err := printer.PrintObj(objToPrint, w); err != nil {
-			if !errs.Has(err.Error()) {
-				errs.Insert(err.Error())
-				allErrs = append(allErrs, err)
-			}
-			continue
-		}
+	convertedList, err := runtime.Decode(legacyscheme.Codecs.UniversalDecoder(), listData)
+	if err != nil {
+		return err
 	}
-	w.Flush()
-	nonEmptyObjCount := 0
-	for _, obj := range objs {
-		if table, ok := obj.(*metav1beta1.Table); ok {
-			// exclude any Table objects with empty rows from our total object count
-			if len(table.Rows) == 0 {
-				continue
-			}
-		}
 
-		nonEmptyObjCount++
+	if err := meta.SetList(convertedList, list.Items); err != nil {
+		return err
 	}
 
-	if nonEmptyObjCount == 0 && !options.IgnoreNotFound {
-		fmt.Fprintln(options.ErrOut, "No resources found.")
+	w := printers.GetNewTabWriter(options.Out)
+	if err := printer.PrintObj(convertedList, w); err != nil {
```

Comment:
@deads2k what do you think if we change the signature of this method to be `PrintObj(out io.Writer, objs ...runtime.Object)` ? So that we can get rid of this nasty hacks above?

Reply:
> @deads2k what do you think if we change the signature of this method to be PrintObj(out io.Writer, objs ...runtime.Object) ? So that we can get rid of this nasty hacks above?

Did the visitor ever get put back in place?  My initial look at this pull found that this problem appeared because it changed a visitor pattern to an aggregation pattern.

Reply:
> PrintObj(out io.Writer, objs ...runtime.Object) ? So that we can get rid of this nasty hacks above?

And I'm against. That is going to encourage people to think they can pass an entire list through and they'll write printers that don't work and attempt to do things like sorting which have to happen above a printer.

Reply:
> sorting which have to happen above a printer.

Maybe I misunderstood you, but I thought that sorting should be part of the printing, should it not?

Reply:
> Did the visitor ever get put back in place? My initial look at this pull found that this problem appeared because it changed a visitor pattern to an aggregation pattern.

We were not visiting here before, we had a slice of runtime.Objects that was passed to the sorter, the slice was sorted in place, and then looped over. For each object, a PrinterForMapping was obtained, and each object was passed to that printer's PrintObj method.

Reply:
If we are doing sorting at the command level, we could keep the slice of runtime.Objects that was here before, sort it in place, then loop over each object and pass it to the printer obtained outside of the loop via `PrintFlags#ToPrinter`, and let the printer keep track internally of the last type that it saw, deciding whether or not to print headers.